### DISCUSSION TITLE
Different results depending on batch size at test time

### DISCUSSION THREAD

**Discussion by aaran2** [pyg-team/pytorch_geometric#4637] [Q&A] [answered]
Hello! First of all thanks for all the great work you guys are doing here :). I'm having a tricky issue (similar to #2665) where the results and the embeddings produced for the same node differ drastically depending on the batch size. I'm doing link prediction on a heterogeneous graph, with the new LinkNeighborLoader. The graph is actually a bipartite graph with two node types (type A and B) and 2 edge types (plus the reverse edges), where each node type only interacts with the other type. The objective is simply to predict the if there is an edge between two nodes. I don't think this is related to the issue but there's somewhat of a temporal component and nodes of type B change across snapshots, and during inference I add a bunch of new edges with new nodes of type B by appending them to the original graph and remapping the IDs, something similar to:

```
inference_graph = graph.clone()
inference_graph["B"].x = torch.cat((graph["B"].x, test_feats["B"]))
# Remap node indices
incoming_edges[1,:] += len(graph["B"].x)
# Only this connection is added since I do not want to update nodes of type A with the new node B during inference
inference_graph["A", "rev_link_2", "B"].edge_index = torch.cat((inference_graph["A", "rev_link _2", "B"].edge_index, incoming_edges), dim=1)
```

I then want to perform link prediction for each of these new nodes, on the other edge type. Let's assume the following scenario:

```
l1 = LinkNeighborLoader(inference_graph, edge_label_index=(("A", "link", "B"),
                  torch.tensor([[717607], [18003125]])),
                  neg_sampling_ratio=0.0, shuffle=False,
                  num_neighbors=[64,64,64],
                  batch_size=1, num_workers=6)


l2 = LinkNeighborLoader(inference_graph, edge_label_index=(("A", "link", "B"),
                  torch.tensor([[717607, 717607], [18003125, 19015616]])),
                  neg_sampling_ratio=0.0, shuffle=False,
                  num_neighbors=[64,64,64],
                  batch_size=2, num_workers=6)

model.eval()
with torch.no_grad():
    for batch in l1:
        batch = batch.to(device)
        logits = model(batch.x_dict, batch.edge_index_dict, batch["A", "link", "B"].edge_label_index
        print(torch.sigmoid(logits))
   
     for batch in l2:
        batch = batch.to(device)
        logits = model(batch.x_dict, batch.edge_index_dict, batch["A", "link", "B"].edge_label_index
        print(torch.sigmoid(logits))
```
With this code we get these results:
![image](https://user-images.githubusercontent.com/105492714/168272749-4020d69b-720c-4f28-89ab-ab5a56af8dc8.png)

The output for edge 717607 -> 18003125 changes from 0.99 to 0.03. The sampled subgraph in both scenarios is identical, since there's less connections than the number of specified neighbors. Furthermore, both edges in l2 have the exact same features and subgraphs, and they get different results. Using a negative_sampling_ratio > 0 also changes the results. In this case, depending on the negative edge sampled, results for the positive edge fluctuate all over. The output embeddings of the very first conv layers are different, so it seems to be an issue outside the model

Any ideas as to what might be happening?

**Comment by Padarn** [comment, marked as answer]
Hey @aaran2 thanks for the question.

Do you mind providing a small example with your model and a dataset to reproduce this? I couldn't reproduce the bug on a very simple model unfortunately.

**Reply by aaran2** [reply]
Hello @Padarn. Some fluctuation can be obtained in the following code (apologies for the bizzare graph, was trying to replicate only the small subgraph described above)
```
class GNN(torch.nn.Module):
    def __init__(self, hidden_channels):
        super().__init__()
        self.convs = torch.nn.ModuleList()
        for i in range(3):
            self.convs.append(HeteroConv({
                    ('A', 'to', 'B'): GATConv((-1, -1), hidden_channels, heads=4, dropout=0.2),
                    ('A', 'rev_to_2', 'B'): GATConv((-1, -1), hidden_channels, heads=4, dropout=0.2),
                    ('B', 'to_2', 'A'): GATConv((-1, -1), hidden_channels, heads=4, dropout=0.2),
                    ('B', 'rev_to', 'A'): GATConv((-1, -1), hidden_channels, heads=4, dropout=0.2),
                }, aggr='sum'))
        
        self.linear = torch.nn.Linear(4*hidden_channels*2, 1)

    def forward(self, x, edge_index, edge_label_index):
        for i in range(len(self.convs)):
            x = self.convs[i](x, edge_index)
            x = {key: x.relu() for key, x in x.items()}
        in_linear = torch.cat((x["A"][edge_label_index[0,:]],
                              x["B"][edge_label_index[1,:]]), dim=-1)
        x = self.linear(in_linear)
        return x
    
device = torch.device('cuda:0') if torch.cuda.is_available() else torch.device('cpu')
t_model = GNN(32).to(device)

graph = HeteroData()
graph["A"].x = torch.rand((4672592, 87))
graph["B"].x = torch.rand((17734764, 34))
graph["A", "to", "B"].edge_index = torch.tensor([[4570844, 4570844, 231501, 231501 , 2024319, 4570844, 4570844],
                                                 [15818902, 16175471, 17397782, 17344776 , 14069160, 15574100, 15163722]])

graph["A", "rev_to_2", "B"].edge_index = torch.tensor([[717607,717607, 4570844, 4570844, 4570844, 4195911, 4195911],
                                                       [15818902, 16175471, 17397782, 17344776, 14069160, 15574100, 15163722]])

graph["B", "to_2", "A"].edge_index = torch.tensor([[15818902, 16175471, 17397782, 17344776, 14069160],
                                                   [717607, 717607, 4570844, 4570844, 4570844]])

graph["B", "rev_to", "A"].edge_index = torch.tensor([[15818902, 16175471, 15574100, 15163722],
                                                     [4570844, 4570844, 4570844, 4570844]])

# Insert some new nodes and corresponding edges into dummy graph
graph["B"].x = torch.cat((graph["B"].x, torch.rand((1,34)).repeat(3926754,1)))
graph["A", "rev_to_2", "B"].edge_index = torch.cat((graph["A", "rev_to_2", "B"].edge_index, 
                                                   torch.tensor([[4570844, 4570844, 3571840, 2196493],
                                                                 [18003125, 19015616, 17734764, 17734765]])), dim=1)

loader_1 = LinkNeighborLoader(graph, edge_label_index=(("A", "to", "B"),
                     torch.tensor([[717607], [18003125]])),
                  neg_sampling_ratio=0.0, shuffle=False,
                  num_neighbors=[32,32,32],
                  batch_size=1, num_workers=4)

loader_2 = LinkNeighborLoader(graph, edge_label_index=(("A", "to", "B"),
                     torch.tensor([[717607, 717607], [18003125, 19015616]])),
                  neg_sampling_ratio=1.0, shuffle=False,
                  num_neighbors=[32,32,32],
                  batch_size=1, num_workers=4)
t_model = t_model.to(device)
t_model.eval()
with torch.no_grad():
    for batch in loader_1:
        batch = batch.to(device)
        logits = t_model(batch.x_dict, batch.edge_index_dict, batch["A", "to", "B"].edge_label_index)
        print(logits)
        break
        
    for batch in loader_2:
        batch = batch.to(device)
        logits = t_model(batch.x_dict, batch.edge_index_dict, batch["A", "to", "B"].edge_label_index)
        print(logits)
        break
```

Running this yields different results for the same edge. Furthermore, running the second for loop multiple times yields different results each time. Fluctuation can also be observed with batch > 1 and neg sampling = 0 in loader_2, although less noticeable. Interestingly, I've noticed that switching GATConv to e.g., SAGEConv resolves the problem, so perhaps it's related to some interaction while calculating attention coefficients?

**Reply by rusty1s** [reply]
Can you try to add `add_self_loops=False` as argument to the `GATConv` calls? This will produce incorrect behavior in bipartite message passing, and may hopefully resolve this issue :)

**Reply by Padarn** [reply]
Good point @rusty1s - do you think we should handle this inside `HetroConv` somehow?

@aaran2 thanks for the example - let me know if @rusty1s's suggestion does not resolve the problem and I'll try to look into it a bit more.

**Reply by rusty1s** [reply]
That's a good suggestion. We could check whether any GNN is used that supports `add_self_loops` and set it to `False` automatically in case `src_node_type != dst_node_type`.

**Reply by Padarn** [reply]
I'll open an issue for it

**Reply by aaran2** [reply]
Hello @rusty1s and @Padarn. Indeed setting `add_self_loops=False` seems to have fixed it, thanks a lot for the help :)!
